File: 1048-longest-string-chain/1048-longest-string-chain.cpp
class Solution {
public:
    static bool f(string &l,string &r) {
        return l.size() < r.size();
    }
    
    int longestStrChain(vector<string>& words) {
        sort(words.begin(),words.end(),f);
        unordered_map<string,int> dp;
        int result = 1;
        string prev;
        for(string w:words) {
            dp[w] = 1;
            for(int i=0;i<w.size();i++) {
                prev = w.substr(0,i) + w.substr(i+1);
                if(dp.find(prev)!=dp.end()) {
                    dp[w] = max(dp[w],1+dp[prev]);
                    result = max(dp[w],result);           
                }
            }
        }
        return result;
    }
};

File: 1048-longest-string-chain/1048-longest-string-chain.py
class Solution:
    def longestStrChain(self, words: List[str]) -> int:
        dp = {}
        words.sort(key=len)
        result = 1
        for w in words:
            dp[w] = 1
            for i in range(len(w)):
                prev = w[:i]+w[i+1:]
                if prev in dp:
                    dp[w] = max(dp[w],dp[prev]+1)
                    result = max(dp[w],result)
        return result